protos/core/Btfs.proto

```diff
@@ -0,0 +1,230 @@
+  	rpc checkFileStoreMeta(CheckFileStoreMetaRequest) returns (FileStoreMeta){} //reserved for other nodes to check contract status
+}
+
+service HostService{
+	rpc challenge(ProofOfReplicateChallenge) returns (ProofOfReplicateChallenge){} //used for receieve challenge question and return the answer, TBD seems can be removed as the BTFS has embed function
+	rpc cancelContract(Contract) returns (Result){} 
+}
+
+//the relationship of the objects from Guard's view
+// FileStoreMeta (1:n) Contract (1:n) payout
+
+
+
+
+
+
+//contract is the signed document between the renter and host for one shard's storage
+// the contract information contains buyer, seller, escrow, guard , storage shard's information, rent duration, amount, payment method(periodically and times)
+message Contract{
+	bytes contract_id = 1; //one id which generated by renter, and will be check unique by escrow service
+	bytes renter_address = 2; //renter's wallet address
+	bytes host_address = 3; //host’s wallet address
+	bytes shard_hash = 4; //shard’s hash
+	int shard_index = 5; //shard’s sequence of the original file
+	int64 shard_file_size = 6; //shard’s file size
+	bytes file_hash = 7; //original file's hash, reserved point back to FileStoreMeta
+	Timestamp rent_start = 7; //rent start time
+	Timestamp rent_end = 8 ; //rent end time
+	bytes guard_address = 9; //renter proposed guard's address
+	bytes escrow_address = 10; //renter proposed escrow's address
+	bytes renter_propose_signature = 11; //renter signature for proposed information, signed for above information
+	int64 agreed_price = 12; //the price which agreed by host and rental, on the first phase, it equal to the proposed_price
+	int64 agreed_amount = 13; //the amount which agreed by host and rental, on the first phase, it should be equal file_size*duration*agreed_price
+	int64 agreed_collaeteral_amount = 14; 
+	enum Schedule{		
+    	MONTHLY = 0;
+    	QUARTERLY = 1;
+    	ANNUALLY = 2;
+	}
+    Schedule payout_schedule = 15; // schedule for payouts
+   	int32 num_payouts = 16; // numbers of payouts
+	Timestamp escrow_signed_time = 17; //contract signed time
+	bytes renter_signature = 18; //renter’s signature for the sub contract
+	bytes host_signature = 19; //host’s signature from the sub contract
+	bytes escrow_signature = 20； //escrow's signature 
+	enum State{
+		DRAFT = 0; //the contract was draft, but the escrow service did not received the money
+		SIGNED = 1;  // the escrow service received the money from renter, the contract was start.
+		UPLOADED = 2; //the first challenge was done, which means the shard was in the host, all pending payouts can be issued now
+		LOST = 3; // the challenge failed, the following payout should be hold or even stopped
+		CANCELED = 4; //if the shard was rebuild after LOST, or after some days of NOQUESTION, it go to CANCELED status
+		CLOSED = 5; // the all work in such contract was done
+	}
+	State state = 21;
```

Comment:
I prefer moving `state` a bit on top of the definition because it determines which fields may be valid during which state. Currently the only downside of this design is that this is not very clear - let's discuss